src/chainstate/stacks/miner.rs

```diff
@@ -326,6 +393,218 @@ impl StacksBlockBuilder {
+                    work: parent_chain_tip.block_height.checked_add(1).expect("FATAL: block height overflow")
+                };
+
+                StacksBlockBuilder::from_parent_pubkey_hash(0, &parent_chain_tip, &new_work, &proof, pubkey_hash)
+            };
+
+        Ok(builder)
+    }
+
+    /// Given access to the mempool, mine an anchored block with no more than the given execution cost.
+    pub fn build_anchored_block(burndb: &mut BurnDB,
+                                chainstate: &mut StacksChainState,
+                                mempool: &MemPoolDB,
+                                parent_burn_header_hash: &BurnchainHeaderHash,
+                                parent_block_hash: &BlockHeaderHash,
+                                proof: VRFProof,
+                                pubkey_hash: Hash160,
+                                coinbase_tx: &StacksTransaction,
+                                execution_budget: ExecutionCost) -> Result<StacksBlock, Error> {
+
+        if let TransactionPayload::Coinbase(..) = coinbase_tx.payload {} else {
+            return Err(Error::MemPoolError("Not a coinbase transaction".to_string()));
+        }
+
+        let mut builder = StacksBlockBuilder::make_block_builder(burndb, chainstate, parent_burn_header_hash, parent_block_hash, proof, pubkey_hash)?;
+
+        builder.set_max_cost(execution_budget.clone());
+        let cost_overflow_recovery_builder = builder.clone();     // used to construct the real anchored block if we run out of budget
+        let mut header_reader_chainstate = chainstate.reopen()?;    // uesd for reading block headers during an epoch
+        
+        let mut tip_height = MemPoolDB::get_block_height(mempool.conn(), parent_burn_header_hash, parent_block_hash)?.unwrap_or(0);
+        let mut tip_burn_header_hash = parent_burn_header_hash.clone();
+        let mut tip_block_hash = parent_block_hash.clone();
+        
+        let mut epoch_tx = builder.epoch_begin(chainstate)?;
+        builder.try_mine_tx(&mut epoch_tx, coinbase_tx)?;
+
+        let mut considered = HashSet::new();
+        let mut next_timestamp = match MemPoolDB::get_next_timestamp(mempool.conn(), parent_burn_header_hash, parent_block_hash, 0)? {
+            Some(ts) => ts,
+            None => {
+                // no transactions at this chain tip, so mine an empty block 
+                let block = builder.mine_anchored_block(&mut epoch_tx);
+                builder.epoch_finish(epoch_tx);
+                return Ok(block);
+            }
+        };
+        
+        test_debug!("Begin scanning transaction mempool at {}/{} height={}", &tip_burn_header_hash, &tip_block_hash, tip_height);
+
+        // set to true if we exceed budget, and need to rebuild with known-good transactions.
+        let mut do_rebuild = false;
+
+        loop {
+            let mut available_txs = match MemPoolDB::get_txs_at(mempool.conn(), &tip_burn_header_hash, &tip_block_hash, next_timestamp) {
```

Comment:
Do we need to query mempool transactions by timestamp? Could we just select all transactions that arrived at a given block burnhash/blockhash here?

Reply:
The timestamp is acting as a "page number" here, in order to avoid an OOM condition where a crap-ton of transactions on a chain tip get loaded into RAM.  Given that there's no bound on how many transactions can exist on a chain tip, and given that the time between anchored blocks can be really long (potentially an hour in some case), I think this OOM condition is plausible.